### PR TITLE
Support queries yielding heterogeneous results

### PR DESCRIPTION
**Pull request by alextwoods**
*Issue:* #107 

*Description of changes:*
Add `#multi_model_filter` to `BuildableSearch` which takes either a passed Proc/lambda or a block.  Given the raw item data that `model_filter` returns the class (or nil to skip) to use for that item.

*Example Usage*
```ruby
class Model_A
  include Aws::Record
  set_table_name(TABLE_NAME)

  string_attr :uuid, hash_key: true
  string_attr :class_name, range_key: true

  string_attr :attr_a
end

class Model_B
  include Aws::Record
  set_table_name(TABLE_NAME)

  string_attr :uuid, hash_key: true
  string_attr :class_name, range_key: true

  string_attr :attr_b
end

# use multi_model_filter to create a query on TABLE_NAME
items = Model_A.build_scan.multi_model_filter do |raw_item_attributes|
  case raw_item_attributes['class_name']
  when "A" then Model_A
  when "B" then Model_B
  else
    nil
  end
end.complete!
```

Of note: 
* This still requires calling `complete!` on the query which is a little strange.  But allows you to pass a lambda/proc was a parameter to `multi_model_filter` and use it just like any other builder function.  (One alternative is to call complete! when a block is passed, but treat a passed parameter as a builder function).
* The item data is raw - which means keys are the raw strings from the db (rather than any translated, symbolized names). 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by awood45 — COMMENTED** [review]
At a glance this looks pretty good. May also want to consider writing integration tests for this.

File: lib/aws-record/record/item_collection.rb [hunk 1]
```diff
@@ -91,9 +92,16 @@ def empty?
...
         items.each do |item|
-          record = model.new
+          model_class = model
+          if @model_filter
+            model_class = @model_filter.call(item)
+            next unless model_class # skip nils
+            record = model_class.new
+          else
+            record = model.new
+          end
```
Comment (sereneiconoclast) [inline lib/aws-record/record/item_collection.rb:102]: A couple alternatives -- if you dislike using ternaries, you could just factor out the `.new`:

```
model_class = model
if @model_filter
  model_class = @model_filter.call(item)
  next unless model_class
end
record = model_class.new
```

If you're fine with ternaries:

```
model_class = @model_filter ? @model_filter.call(item) : model
next unless model_class
record = model_class.new
```

(The latter assumes the supplied `model` is not `nil`. But it looks like `model` is always `@model`. So you could prevent that case by checking for it in the constructor.)

**Comment by alextwoods** [comment]
Was out last week - but will pick this back up this week and add integ tests.

File: lib/aws-record/record/item_collection.rb (hunk 1 above)
Comment (alextwoods) [reply lib/aws-record/record/item_collection.rb]: I think the ternary here is concise and clear.  I've cleaned up following this - good suggestion.

File: features/step_definitions.rb [hunk 2]
```diff
@@ -17,15 +17,15 @@
 
 def cleanup_table
   begin
-    log "Cleaning Up Table: #{@table_name}"
```
Comment (mullermp) [inline features/step_definitions.rb:20]: newer cucumber says to use log

Comment (alextwoods) [reply features/step_definitions.rb]: Switched back to log.

**Comment by alextwoods** [comment]
Have had to prioritize other work - but returning to this now.  I will merge this tomorrow unless there are any other comments.

**Pull request merged**